### PR TITLE
Carrie perception

### PR DESCRIPTION
**Pull request by laibazahid26**
### Add SAM3 instance segmentation perception pipeline

This PR consists of two ROS2 packages: 
**1. carrie_interfaces:** 
custom service definition called as `DetectObjects.srv`. This definition is used by the service server and service client in package `carrie_perception`.

Service accepts Image + prompts, returns Detection2DArray + mask

**2. carrie_perception**
contains the client that makes a segmentation request and contains a sever that segments an Image based on a text prompt.

### CODE REVIEW

**Comment by laibazahid26** [comment]
Please tell me if the weights take forever to download. I came across this: https://discuss.huggingface.co/t/download-speed-way-too-slow/169824/4
I can make use of the info here. 

For me, it took 46 min and 17 sec to download, which isn't too bad I believe? 

image

**Comment by laibazahid26** [comment]
update: changed the architecture as per our discussion. Instead of publishers and subscribers we have a service based architecture.

File: carrie_perception/carrie_perception/helper_functions.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
+import supervision as sv
```
Comment (griswaldbrooks) [inline carrie_perception/carrie_perception/helper_functions.py]: Sounds like this whole file should be deleted

File: carrie_interfaces/srv/DetectObjects.srv [hunk 2]
```diff
@@ -0,0 +1,7 @@
+sensor_msgs/Image image
+string prompt
+---
+vision_msgs/Detection2DArray detections
```
Comment (griswaldbrooks) [inline carrie_interfaces/srv/DetectObjects.srv]: I don't think this is the right move here. Looking at the message definition, this makes more sense when  the model itself is coming up with the classes, but in this case the caller of the service already has the prompt and so there isn't new information here. At best, you should add the prompt as a string back to the response message but that's it. And add a header so this message can be timestamped.

File: carrie_interfaces/srv/DetectObjects.srv [hunk 3]
```diff
@@ -0,0 +1,7 @@
...
+sensor_msgs/Image mask
```
Comment (griswaldbrooks) [inline carrie_interfaces/srv/DetectObjects.srv]: We don't want to merge the masks into a single mask, we should return the masks separately. They will be hard to break up again if we want to do further processing on them if they are combined. This should be an array.

File: carrie_interfaces/srv/DetectObjects.srv [hunk 4]
```diff
@@ -0,0 +1,7 @@
...
+bool success
+string message
```
Comment (griswaldbrooks) [inline carrie_interfaces/srv/DetectObjects.srv]: Make this `error_message`. I don't think there's value in having an arbitrary "7 detections!" or something like that since we should be passing those back as separate masks that can be counted.

File: carrie_interfaces/srv/DetectObjects.srv (hunk 2 above)
Comment (griswaldbrooks) [reply carrie_interfaces/srv/DetectObjects.srv]: And confidence scores. Those should be returned too.

File: carrie_perception/carrie_perception/sam3_service_server.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+class Sam3Detector(Node):
+    def __init__(self):
+        super().__init__('sam3_detector')
+
+        # parameters
+        self.declare_parameters(
+            namespace='',
+            parameters=[
+                ('threshold', 0.5),
```
Comment (griswaldbrooks) [inline carrie_perception/carrie_perception/sam3_service_server.py]: Can we name this `confidence_threshold`?

File: carrie_perception/carrie_perception/sam3_service_server.py [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+        # load model
+        self.model = Sam3Model.from_pretrained("facebook/sam3").to(self.device)
+        self.processor = Sam3Processor.from_pretrained("facebook/sam3")
+
+        self.bridge = CvBridge()
+
+        self.srv = self.create_service(DetectObjects, 'detect_objects', self.detect_callback)
+
+        self.output_saved = False
+        os.makedirs("carrie_perception/sam3_output", exist_ok=True)
```
Comment (griswaldbrooks) [inline carrie_perception/carrie_perception/sam3_service_server.py]:
```suggestion
```

File: carrie_perception/carrie_perception/sam3_service_server.py [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+            parameters=[
+                ('threshold', 0.5),
+                ('mask_threshold', 0.5),
+            ])
+
+        self.threshold = self.get_parameter('threshold').value
+        self.mask_threshold = self.get_parameter('mask_threshold').value
+
+        # device
+        self.device = torch.device('mps' if torch.backends.mps.is_available() else 'cpu')
```
Comment (griswaldbrooks) [inline carrie_perception/carrie_perception/sam3_service_server.py]: The device selection only supports MPS and CPU. Can we add CUDA support as well?
```python
if torch.cuda.is_available():
      self.device = torch.device('cuda')
elif torch.backends.mps.is_available():
      self.device = torch.device('mps')
else:
      self.device = torch.device('cpu')
```

File: carrie_perception/carrie_perception/sam3_service_server.py [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+
+        self.srv = self.create_service(DetectObjects, 'detect_objects', self.detect_callback)
+
+        self.output_saved = False
+        os.makedirs("carrie_perception/sam3_output", exist_ok=True)
+
+        self.get_logger().info('SAM3 detector service ready')
+
+    def detect_callback(self, request: DetectObjects.Request, response: DetectObjects.Response):
+        prompt = request.prompt if request.prompt else 'person'
```
Comment (griswaldbrooks) [inline carrie_perception/carrie_perception/sam3_service_server.py]:
```suggestion
```

File: carrie_perception/carrie_perception/sam3_service_server.py [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+
+        try:
+            # convert ROS image -> PIL
+            cv_image = self.bridge.imgmsg_to_cv2(request.image, desired_encoding='rgb8')
+            pil_image = PILImage.fromarray(cv_image)
```
Comment (griswaldbrooks) [inline carrie_perception/carrie_perception/sam3_service_server.py]: The Sam3Processor accepts numpy arrays directly (which is what opencv images are in python), so the PIL conversion should be unnecessary. You could do:
```python
  cv_image = self.bridge.imgmsg_to_cv2(request.image, desired_encoding='rgb8')

  inputs = self.processor(
      images=cv_image,
      text=prompt,
      return_tensors='pt'
  ).to(self.device)
```

File: carrie_perception/carrie_perception/sam3_service_server.py [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+
+            results = self.processor.post_process_instance_segmentation(
+                outputs,
+                threshold=self.threshold,
+                mask_threshold=self.mask_threshold,
+                target_sizes=[pil_image.size[::-1]]
+            )[0]
+
+            # build Detection2DArray
+            detections_msg = Detection2DArray()
```
Comment (griswaldbrooks) [inline carrie_perception/carrie_perception/sam3_service_server.py]: See my earlier comment about message response

File: carrie_perception/carrie_perception/sam3_service_server.py [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+            for i, score in enumerate(results["scores"]):
+                det = Detection2D()
+                det.header = request.image.header
+
+                box = results["boxes"][i]
+
+                cx = float((box[0] + box[2]) / 2.0)
+                cy = float((box[1] + box[3]) / 2.0)
+                w  = float(box[2] - box[0])
+                h  = float(box[3] - box[1])
```
Comment (griswaldbrooks) [inline carrie_perception/carrie_perception/sam3_service_server.py]: Things like this should be standalone functions

File: carrie_perception/carrie_perception/sam3_service_server.py [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+                det.bbox.size_y = h
+
+                hyp = ObjectHypothesisWithPose()
+                hyp.hypothesis.class_id = prompt
+                hyp.hypothesis.score = float(score)
+                det.results.append(hyp)
+
+                detections_msg.detections.append(det)
+
+            # build combined mask
```
Comment (griswaldbrooks) [inline carrie_perception/carrie_perception/sam3_service_server.py]: We should return individual masks

File: carrie_perception/carrie_perception/sam3_service_server.py [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+
+                detections = helper_functions.from_sam(sam_result=sam_result_for_helper)
+                detections = detections[detections.confidence > self.threshold]
+
+                if len(detections) > 0:
+                    annotated = helper_functions.annotate(pil_image, detections, label=prompt)
+                    output_path = os.path.join("carrie_perception/sam3_output", "annotated_output.jpg")
+                    annotated.save(output_path)
+                    self.get_logger().info(f"Saved annotated debug image to {output_path}")
+                    self.output_saved = True
```
Comment (griswaldbrooks) [inline carrie_perception/carrie_perception/sam3_service_server.py]:
```suggestion
```

File: carrie_perception/carrie_perception/sam3_service_server.py (hunk 8 above)
Comment (griswaldbrooks) [reply carrie_perception/carrie_perception/sam3_service_server.py]: If there's no prompt we should return

File: carrie_perception/carrie_perception/test_service_client.py [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+        self.bridge = CvBridge()
+        self.client = self.create_client(DetectObjects, 'detect_objects')
+
+        # load image
+        if not os.path.exists(self.image_path):
+            self.get_logger().error(f'Image not found: {self.image_path}')
+            rclpy.shutdown()
+            return
+
+        pil_image = PILImage.open(self.image_path).convert('RGB')
```
Comment (griswaldbrooks) [inline carrie_perception/carrie_perception/test_service_client.py]: You don't need PIL. Opencv can open images.

**Review by griswaldbrooks — CHANGES_REQUESTED** [review]

**Comment by griswaldbrooks** [comment]
It'd be great to have a unit test for `detect_callback` that doesn't require loading the real model or spinning up ROS. Here's an example using `unittest.mock.patch`:                         
```python                                                                                                                                                                                             
  import unittest                                                                                                                                                                            
  from unittest.mock import patch, MagicMock                                                                                                                                                 
  import numpy as np                                                         
  import torch
  import rclpy
  from sensor_msgs.msg import Image
  from std_msgs.msg import Header
  from carrie_interfaces.srv import DetectObjects


  class TestSam3Detector(unittest.TestCase):
      @classmethod
      def setUpClass(cls):
          rclpy.init()

      @classmethod
      def tearDownClass(cls):
          rclpy.shutdown()

      @patch('carrie_perception.sam3_service_server.os.makedirs')
      @patch('carrie_perception.sam3_service_server.Sam3Processor')
      @patch('carrie_perception.sam3_service_server.Sam3Model')
      def test_detect_callback_returns_detections(self, MockModel, MockProcessor, _mock_makedirs):
          # Set up fake model that returns a mock output
          mock_model = MagicMock()
          MockModel.from_pretrained.return_value.to.return_value = mock_model

          # Set up fake processor that returns fake post-processed results
          mock_processor = MagicMock()
          MockProcessor.from_pretrained.return_value = mock_processor
          mock_processor.return_value.to.return_value = MagicMock()  # preprocessed inputs
          mock_processor.post_process_instance_segmentation.return_value = [{
              "scores": torch.tensor([0.9, 0.8]),
              "boxes": torch.tensor([[10., 20., 50., 60.], [100., 200., 150., 250.]]),
              "masks": torch.ones((2, 480, 640), dtype=torch.int64),
          }]

          from carrie_perception.sam3_service_server import Sam3Detector
          node = Sam3Detector()

          # Build a fake request
          request = DetectObjects.Request()
          request.image = Image()
          request.image.header = Header()
          request.image.height = 480
          request.image.width = 640
          request.image.encoding = 'rgb8'
          request.image.step = 640 * 3
          request.image.data = bytes(np.zeros((480, 640, 3), dtype=np.uint8))
          request.prompt = 'person'

          response = DetectObjects.Response()
          response = node.detect_callback(request, response)

          assert response.success
          assert len(response.detections.detections) == 2
          assert response.detections.detections[0].bbox.center.position.x == 30.0  # (10+50)/2
          assert response.detections.detections[0].bbox.size_x == 40.0  # 50-10

          node.destroy_node()
```
  The key idea: monkey patch `Sam3Model` and `Sam3Processor` at the module level so the real weights never load, then call detect_callback directly with a synthetic request. ros2 is initialized minimally with rclpy.init() just so the node can be constructed, but in other situations we can even patch out ros.

**Comment by griswaldbrooks** [comment]
Additionally, it would be best if much of your node was composed of named functions that can be unit tested independently and the above test example was just ensuring the whole thing worked together.

```python
# Standalone, unit-testable functions
  def xyxy_to_cwh(box):
      ...
      return cx, cy, w, h


  def masks_to_images(masks):
      """Convert model mask tensor to list of individual uint8 numpy arrays."""
      ...
      return ...
```
Then callback becomes glue code
```python
  def detect_callback(self, request, response):
      try:
          cv_image = self.bridge.imgmsg_to_cv2(request.image, desired_encoding='rgb8')

          inputs = self.processor(...

          with torch.no_grad():
              outputs = self.model(**inputs)

          results = self.processor.post_process_instance_segmentation(...

          mask_images = masks_to_images(results["masks"])
          response.masks = [...
          ...
          response.boxes = []
          for box in results["boxes"]:
              cx, cy, w, h = xyxy_to_cwh(box)
              ...
      ...
      return response
```
Then it can be unit tested without ros or sam
```python
  def test_xyxy_to_cwh():
      assert xyxy_to_cwh([10., 20., 50., 60.]) == (30., 40., 40., 40.)

  def test_masks_to_images():
      masks = torch.ones((3, 480, 640), dtype=torch.int64)
      result = masks_to_images(masks)
      assert len(result) == 3
      assert result[0].shape == (480, 640)
      assert result[0].max() == 255

  def test_masks_to_images_empty():
      masks = torch.zeros((0, 480, 640), dtype=torch.int64)
      result = masks_to_images(masks)
      assert len(result) == 0
```

File: carrie_perception/carrie_perception/test_sam3_service_server.py [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+
+
+class TestXyxyToCwh(unittest.TestCase):
+
+    def test_basic_conversion(self):
+        cx, cy, w, h = xyxy_to_cwh([10., 20., 50., 60.])
+        self.assertEqual(cx, 30.0)
+        self.assertEqual(cy, 40.0)
+        self.assertEqual(w,  40.0)
+        self.assertEqual(h,  40.0)
```
Comment (griswaldbrooks) [inline carrie_perception/carrie_perception/test_sam3_service_server.py]:
```suggestion
        # GIVEN two corners of the image
        # set the variables
        # WHEN converting the center width height
        cx, cy, w, h = xyxy_to_cwh([10., 20., 50., 60.])
        # THEN we expect them to be shifted appropriately
        self.assertEqual(cx, 30.0)
        self.assertEqual(cy, 40.0)
        self.assertEqual(w,  40.0)
        self.assertEqual(h,  40.0)
```
